includes/server/class-llms-rest-memberships-controller.php

```diff
@@ -0,0 +1,590 @@
+				'- `none` displays the post content. ' .
+				'- `content` displays alternate content from the `excerpt` property.' .
+				'- `page` redirects to the WordPress page defined in `content_page_id`.' .
+				'- `url` redirects to the URL defined in `content_page_url`.',
+				'lifterlms' ),
+			'type'        => 'string',
+			'default'     => 'none',
+			'enum'        => array_keys( llms_get_sales_page_types() ),
+			'context'     => array( 'view', 'edit' ),
+		);
+		$schema['properties']['sales_page_url']      = array(
+			'description' => __(
+				'The URL of the sales page content. Required when `sales_page_type` equals `url`. ' .
+				'Only returned when the `sales_page_type` equals `url`.',
+				'lifterlms' ),
+			'type'        => 'string',
+			'context'     => array( 'view', 'edit' ),
+			'format'      => 'uri',
+			'arg_options' => array(
+				'sanitize_callback' => 'esc_url_raw',
+			),
+		);
+		$schema['properties']['tags']                = array(
+			'description' => __( 'List of membership tags.', 'lifterlms' ),
+			'type'        => 'array',
+			'items'       => array(
+				'type' => 'integer',
+			),
+			'context'     => array( 'view', 'edit' ),
+		);
+
+		return $schema;
+	}
+
+	/**
+	 * Get object.
+	 *
+	 * @since [version]
+	 *
+	 * @param int $id Object ID.
+	 * @return LLMS_Membership|WP_Error
+	 */
+	protected function get_object( $id ) {
+		/** @var LLMS_Membership $membership */
+		$membership = llms_get_post( $id );
+
+		return is_a( $membership, 'LLMS_Membership' ) ? $membership : llms_rest_not_found_error();
+	}
+
+	/**
+	 * Retrieve an ID from the object.
+	 *
+	 * @since [version]
+	 *
+	 * @param LLMS_Membership $object LLMS_Membership object.
+	 * @return int
+	 */
+	protected function get_object_id( $object ) {
+		return $object->get( 'id' );
+	}
+
+	/**
+	 * Maps a taxonomy name to the relative rest base.
+	 *
+	 * @since [version]
+	 *
+	 * @param object $taxonomy The taxonomy object.
+	 * @return string The taxonomy rest base.
+	 */
+	protected function get_taxonomy_rest_base( $taxonomy ) {
+		$base = ! empty( $taxonomy->rest_base ) ? $taxonomy->rest_base : $taxonomy->name;
+
+		$taxonomy_base_map = array(
+			'membership_cat' => 'categories',
+			'membership_tag' => 'tags',
+		);
+
+		return isset( $taxonomy_base_map[ $base ] ) ? $taxonomy_base_map[ $base ] : $base;
+	}
+
+	/**
+	 * Prepares a single post for create or update.
+	 *
+	 * @since [version]
+	 *
+	 * @param WP_REST_Request $request Request object.
+	 * @return array|WP_Error Array of llms post args or WP_Error.
+	 */
+	protected function prepare_item_for_database( $request ) {
+		$prepared_item = parent::prepare_item_for_database( $request );
+		$schema        = $this->get_item_schema();
+
+		// Restriction action.
+		if ( ! empty( $schema['properties']['restriction_action'] ) && isset( $request['restriction_action'] ) ) {
+			$prepared_item['restriction_redirect_type'] = $request['restriction_action'];
+		}
+
+		// Restriction message.
```

Comment:
See my comment in `update_additional_object_fields()` that will implicitly explain the reason why this should be moved there.